rest_server/IOMgrConnection.go

```diff
@@ -0,0 +1,393 @@
+package main
+
+import (
+	"SiodbIomgrProtocol"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net/http"
+	"unicode/utf8"
+
+	"github.com/gin-gonic/gin"
+	"github.com/golang/protobuf/proto"
+)
+
+var ()
+var (
+	IOMgrChunkMaxBufferedSize uint32 = 1024
+	JSONPayloadBufferSize     uint64 = 64 * 1024
+	MessageLengthMaxSize      uint32 = 1 * 1024 * 1024
+)
+
+var (
+	DATABASEENGINERESPONSE    uint32 = 4
+	DATABASEENGINERESTREQUEST uint32 = 13
+)
+
+type IOMgrConnection struct {
+	*TrackedNetConn
+	pool *IOMgrConnPool
+}
+
+func (IOMgrConn *IOMgrConnection) cleanupTCPConn() (err error) {
+
+	var IOMgrChunkSize uint32
+	var counter int16
+
+	for {
+		// Get Current Row Size
+		if _, IOMgrChunkSize, err = IOMgrConn.readVarint32(); err != nil {
+			return fmt.Errorf("cleanupTCPConn: unable to read chunk size")
+		}
+		if IOMgrChunkSize == 0 {
+			siodbLoggerPool.Debug("cleanupTCPConn: %d chunks dropped.", counter)
+			return err
+		}
+		siodbLoggerPool.Debug("cleanupTCPConn: Chunk size to drop: %d.", IOMgrChunkSize)
+		buff := make([]byte, IOMgrChunkSize)
+		_, err = io.ReadFull(IOMgrConn, buff)
+
+		counter++
+	}
+}
+
+func (IOMgrConn *IOMgrConnection) writeJSONPayload(requestID uint64, c *gin.Context) (err error) {
+
+	// Read Payload up to max json payload size
+	var bytesReadTotal uint64 = 0
+	var body []byte
+	siodbLoggerPool.Debug("IOMgrConn.pool.maxJsonPayloadSize: %v", IOMgrConn.pool.maxJsonPayloadSize)
+
+	for true {
+		buffer := make([]byte, JSONPayloadBufferSize)
+		if JSONPayloadBufferSize > IOMgrConn.pool.maxJsonPayloadSize {
+			buffer = make([]byte, IOMgrConn.pool.maxJsonPayloadSize)
+		}
+		if bytesReadTotal > IOMgrConn.pool.maxJsonPayloadSize {
+			break
+		}
+		bytesRead, err := io.ReadFull(c.Request.Body, buffer)
+		if err == io.ErrUnexpectedEOF {
+			body = AppendBytes(body, buffer[:bytesRead])
```

Comment:
That's bad approach. You potentially allowing to send you huge payload, you accumulate that it memory and then it will be out of memory and crash and out of service as result.
What needs to be done:
1. Drop away that "body"
2. Make sure that bufferSize it is does not exceed 2^31 - 1
3. In the loop, until whole payload is read from client:
  - Receive chunk of payload from HTTP client up to bufferSize
  - Immediately send it to iomgr.
4. Send zero byte to iomgr to indicate that chunk series finished.

Reply:
@cxxman indeed much more simple and cleaner